### PR TITLE
Recover from errors during migration.

### LINKED ISSUE #278: 'TD_Database:openWithEncryptionKeyProvider:': Rework this method so we can rescue failures during migration
Although we exit the method when there is an error while updating the version, on other cases we update the database but we do not check the result, for example:

```objc
            while ([attachmentKeys next]) {
                NSData *keyData = [attachmentKeys dataNoCopyForColumn:@"key"];
                
                [TD_Database generateAndInsertFilenameBasedOnKey:*(TDBlobKey *)keyData.bytes
                                intoBlobFilenamesTableInDatabase:db];
            }
```

### PR DESCRIPTION
**Pull request by rhyshort**
## What
Recover from errors during database migration by make schema updates in
a transaction and rolling back whenever an error occurs.

## How
- Move migrations to being run inside a transaction, so they can be rolled back in the event of an error.
- Check the return value from the `generateAndInsertFilenameBasedOnKey:intoBlobFilenamesTableInDatabase` method, so we can rollback if a attachment failed to be moved.

## Tests

All current tests pass, I'm not entirely sure how to force FMDB to error so we can test rollbacks, any advice on this would be appreciated.

## Issues

Fixes #278

### CODE REVIEW

File: CDTDatastore/touchdb/TD_Database.m [hunk 1]
```diff
@@ -331,9 +331,11 @@ - (BOOL)openWithEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
 
     __block BOOL result = YES;
     __weak TD_Database* weakSelf = self;
-    [_fmdbQueue inDatabase:^(FMDatabase* db) {
+    [self.fmdbQueue inTransaction:^(FMDatabase* db, BOOL* rollback) {
         TD_Database* strongSelf = weakSelf;
 
+        *rollback = YES;
```
Comment (ricellis) [inline CDTDatastore/touchdb/TD_Database.m]: Why not add a comment here that the transaction will be rolled back unless the entire block completes successfully when rollback will be set to NO?
I think that will give us a nice compromise between having all the extra ` = YES` lines and being clear about what is going to happen with the transaction.

File: CDTDatastore/touchdb/TD_Database.m [hunk 2]
```diff
@@ -528,6 +535,8 @@ INSERT INTO INFO (key, value) VALUES ('publicUUID',  '%@')",
             result = NO;
             return;
         }
+
+        *rollback = NO;
```
Comment (ricellis) [inline CDTDatastore/touchdb/TD_Database.m]: Add comment something like `// Block completed successfully, no need to rollback`

Comment (rhyshort) [inline CDTDatastore/touchdb/TD_Database.m]: done in 8336484

File: CDTDatastore/touchdb/TD_Database.m (hunk 1 above)
Comment (rhyshort) [inline CDTDatastore/touchdb/TD_Database.m]: done in 8336484

**Comment by ricellis** [comment]
+1

**Pull request merged**